react-backend/client/src/components/visitors/qrcode.js

```diff
@@ -131,23 +161,36 @@ export default class QRCode extends Component {
       );
     }
     return (
-      <div className="row">
-        <section className="Main col-9">
-          <h1 className="capitalise" id="username">
-              Welcome Back, {this.state.username}
-          </h1>
-
-          {this.state.activities.map(activity => (
-            <PurposeButton
-              key={activity.name}
-              session={activity.name}
-              activity={this.state.activity}
-              onClick={this.changeActivity}
-            />
-          ))}
-        </section>
-        <QRPrivacy className="col-3" />
-      </div>
+      <section>
+        <StyledNav>
+          <Heading>
+              Welcome back, {capitaliseFirstName(this.state.username)}! Why are you here today?
+          </Heading>
+        </StyledNav>
+        <StyledSection>
+          <HalfFlexContainerRow>
+            {this.state.activities.map((activity, index) => (
```

Comment:
I think the UI semantic we're actually trying to express here is "row of buttons with alternating styles".

I think a better way to structure that would be to have an `AlternatingComponentRow`, which would accept as props the two components to render in alternating fashion, as well as a data array which would be passed as props to those two components.

```js
<AlternatingComponentRow 
  one={PrimaryButton}
  two={SecondaryButton}
  data={[ { content: 'Violin', onClick: () => {}, ... }, { content: 'Cafe', ... }, ... ]}
/>
```
Then then all the messy logic can be internal to that.

Reply:
The way I have done this the buttons are always created in the same order, but use flex to switch their ordering when displayed, rather than creating each 2nd row in reveres. The benefit of this is on smaller screen sizes the buttons will appear in the correct order if stacked in 1 column.

Agree that there's probably a better way of writing this. HMU next time you're working I'd find it useful to chat through it.

Reply:
Yeah ok I don't want this changed at the moment, but we can talk about it in future.